File: sshiny/src/jlo/ioe/Sheet.scala
package jlo.ioe;

import jlo.ioe.ui._
import jlo.ioe.ui.Splitter._
import javax.swing.border.{CompoundBorder,LineBorder,EmptyBorder}
import java.awt.Color
import java.awt.event._
import java.io.{ObjectOutput,ObjectInput}

case class SheetTitleChanged(newTitle:String) extends ObservableEvent

object SheetStateStorage extends data.DOStorage[SheetState] {
  val _db = createDB("SheetState", classOf[SheetState])
  def db = _db
}

@SerialVersionUID(1000)
class SheetState(var root : Split[data.Ref[data.DataObject]]) extends data.DataObject { 
  def splits = root

  def this() = this(null)

  def storage = SheetStateStorage
  def kind = "SheetState"
  def defaultView = null

  override def writeExternal(out:ObjectOutput) : Unit = {
    super.writeExternal(out)
    out.writeObject(root)
  }
  override def readExternal(in:ObjectInput) : Unit = {
    super.readExternal(in)
    root = in.readObject.asInstanceOf[Split[data.Ref[data.DataObject]]]
  }
}

class ViewAccessor extends ComponentAccessor[data.Ref[data.DataObject]]({(a)=>null}) {
  override def apply(p : data.Ref[data.DataObject]) : Component = p().defaultView
}

class Sheet(screen:Screen, sstate : SheetState) extends Panel with Observer {
  type SDO = data.Ref[data.DataObject]
  var obj = sstate.splits().get.apply()
  var titleString : String = obj.toString // todo: how to track this?
  var splitPane = new Splitter
  setBorder(new EmptyBorder(2,2,2,2))
  add(splitPane)
  addComponentListener(new ComponentAdapter {
    var firstTime = true
    override def componentShown(e:ComponentEvent) : Unit = {
      if (firstTime) {
	splitPane.resplit(sstate.splits)
	listenTo(obj) event {
	  case data.DataObjectModified(data.MetadataChanged(data.DataObjects.kTitle)) => title = obj.meta(data.DataObjects.kTitle).get("").asInstanceOf[String]
	  case _ => {} 
	}
	firstTime = false
      }
    }
  })

  def this(_screen : Screen, obj : data.DataObject) = { this(_screen, new SheetState(new Split(Some(new data.Ref[data.DataObject](Some(obj))), 
 											       Aspect.Horizontal, 
 											       1.0, 
 											       new ViewAccessor))); this.obj = obj }

  def state = sstate
  override def equals(o:Any) : boolean = {
    if (o.isInstanceOf[Sheet]) {
      val s = o.asInstanceOf[Sheet]
      s.state.objectID == sstate.objectID
    } else false
  }

  def title = titleString
  def title_=(s:String) = {
    titleString = s
    fire(SheetTitleChanged(titleString))
  }
  def display = {
    Console.println("Shown!")
    screen.display(this)
    this
  }

  def split(a:data.DataObject) = {    
    val max = findSplit(sstate.splits,sstate.splits)
    Console.println(max)
    max.aspect match {
      case Aspect.Horizontal() => max.verticalDivider(max().get,new data.Ref[data.DataObject](Some(a)))
      case Aspect.Vertical() => max.horizontalDivider(max().get,new data.Ref[data.DataObject](Some(a)))
    }
    sstate.save
    splitPane.resplit(sstate.splits)
  }
  
  private def findSplit(max:Split[SDO], s:Split[SDO]) : Split[SDO] = {
    Console.println("MAX: " + max)
    Console.println("s  : " + s)
    // add the object at the no split with the largest area
    s.kind match {
      case SplitType.NoSplit() => if (s.area >= max.area) s else max
      case _ => Util.argmax({(a:Split[data.Ref[data.DataObject]],b:Split[data.Ref[data.DataObject]]) => if (a.area > b.area) a else b}, List(findSplit(max,s.first.get),findSplit(max,s.second.get)))
    }
  }
}

File: trunk/jshiny/src/jlo/ioe/Sheet.java
package jlo.ioe;

import jlo.ioe.data.DataObject;
import jlo.ioe.data.Ref;
import jlo.ioe.messaging.AbstractMessageWithParams;
import jlo.ioe.messaging.MessageService;
import jlo.ioe.messaging.SubscriberDelegate;
import jlo.ioe.ui.Aspect;
import jlo.ioe.ui.IComponent;
import jlo.ioe.ui.Panel;
import jlo.ioe.ui.Split;
import jlo.ioe.ui.Splitter;
import jlo.ioe.util.F;
import jlo.ioe.util.Identifiable;
import jlo.ioe.util.ObjectID;
import jlo.ioe.util.Opt;
import jlo.ioe.util.Tuple;
import jlo.ioe.util.Util;

import javax.swing.border.EmptyBorder;
import java.util.Arrays;

/**
 * Copyright © 2007 imaginaryday.com (jlo)<br>
 * User: jlowens<br>
 * Date: Apr 7, 2007<br>
 * Time: 2:25:30 PM<br>
 */
public class Sheet extends Panel implements Identifiable {
	public static Class TitleChanged = TitleChangedMsg.class;

	class TitleChangedMsg extends AbstractMessageWithParams {
		public TitleChangedMsg(Object sender, Tuple.One<String> title) {
			super(sender, title);
		}
	}

	public String getTitle() {
		return title;
	}

	private void setTitle(String title) {
		this.title = title;
		MessageService.singleton().publish(new TitleChangedMsg(Sheet.this, Tuple.one(title)));
	}

	private String title;
	private ObjectID oid = new ObjectID(getClass());
	private Screen screen;
	private DataObject obj;
	private Split<Ref<DataObject>> splits;
	private Splitter splitPane;
	private SubscriberDelegate subscriber = new SubscriberDelegate();


	public Sheet(Screen screen, final DataObject obj) {
		this.screen = screen;
		this.obj = obj;
		splits = new Split<Ref<DataObject>>(Opt.some(new Ref<DataObject>(obj)), Aspect.Horizontal, 1.0, new F.lambda1<IComponent, Ref<DataObject>>(){protected View code(Ref<DataObject> p) {
			return p.get().getDefaultView();
		}});
		splitPane = new Splitter();
		add(splitPane);
		splitPane.resplit(splits);
		setBorder(new EmptyBorder(2,2,2,2));
		subscriber.subscribe(
				new Ref<DataObject>(obj),
				DataObject.Modified,
				DataObject.kTitle,
				new F.lambda0<Object>(){protected Object code() {
					setTitle(obj.meta(DataObject.kTitle).get(""));
					return null;
				}});
	}

	public Sheet display() {
		screen.display(this);
		return this;
	}

	public void split(DataObject obj) {
		Split<Ref<DataObject>> max = findSplit(splits, splits);
		switch (max.aspect) {
			case Horizontal:
				max.verticalDivider(max.obj.get(null), new Ref<DataObject>(obj));
				break;
			case Vertical:
				max.horizontalDivider(max.obj.get(null), new Ref<DataObject>(obj));
				break;
		}
		splitPane.resplit(splits);
	}

	private Split<Ref<DataObject>> findSplit(Split<Ref<DataObject>> max, Split<Ref<DataObject>> s) {
		switch (s.kind) {
			case NoSplit: return (s.area() >= max.area()) ? s : max;
			default:
				return Util.argmax(new F.lambda2<Split<Ref<DataObject>>,Split<Ref<DataObject>>,Split<Ref<DataObject>>>(){protected Split<Ref<DataObject>> code(Split<Ref<DataObject>> t, Split<Ref<DataObject>> u) {
					return (t.area() > u.area()) ? t : u;
				}}, Arrays.asList(findSplit(max, s.first.get(null)), findSplit(max, s.second.get(null))));
		}
	}

	public ObjectID getObjectID() {
		return oid;
	}
}
